web-common/src/features/alerts/alert-preview-data.ts

```diff
@@ -0,0 +1,167 @@
+  return derived(
+    [
+      ctx.runtime,
+      ctx.metricsViewName,
+      ctx.dashboardStore,
+      useMetricsView(ctx),
+      ctx.selectors.timeRangeSelectors.timeControlsState,
+      measureFilterResolutionsStore(ctx),
+    ],
+    (
+      [
+        runtime,
+        metricsViewName,
+        dashboard,
+        metricsView,
+        timeControls,
+        resolvedMeasureFilters,
+      ],
+      set,
+    ) => {
+      return createQueryServiceMetricsViewAggregation(
+        runtime.instanceId,
+        metricsViewName,
+        getAlertPreviewQueryRequest(
+          params,
+          dashboard,
+          metricsView?.data,
+          timeControls,
+          resolvedMeasureFilters,
+        ),
+        {
+          query: getAlertPreviewQueryOptions(
+            ctx,
+            params,
+            metricsView?.data ?? {},
+          ),
+        },
+      ).subscribe(set);
+    },
+  );
+}
+
+function getAlertPreviewQueryRequest(
+  params: AlertPreviewParams,
+  dashboard: MetricsExplorerEntity,
+  metricsViewSpec: V1MetricsViewSpec | undefined,
+  timeControls: TimeControlState,
+  resolvedMeasureFilters: ResolvedMeasureFilter,
+): V1MetricsViewAggregationRequest {
+  const dimensions: V1MetricsViewAggregationDimension[] = [];
+  const sort: V1MetricsViewAggregationSort[] = [];
+
+  const pivotByTime =
+    !!params.splitByTimeGrain && !!metricsViewSpec?.timeDimension;
+  const grain = mapDurationToGrain(params.splitByTimeGrain ?? "");
+  if (pivotByTime && grain !== V1TimeGrain.TIME_GRAIN_UNSPECIFIED) {
+    dimensions.push({
+      name: metricsViewSpec.timeDimension,
+      timeZone: dashboard.selectedTimezone,
+      timeGrain: grain,
+    });
+    sort.push({ name: metricsViewSpec.timeDimension, desc: true });
+  }
+  if (params.dimension) {
+    dimensions.push({ name: params.dimension });
+    sort.push({ name: params.dimension, desc: true });
+  }
+
+  return {
+    measures: [{ name: params.measure }],
+    dimensions,
+    where: sanitiseExpression(
+      dashboard.whereFilter,
+      resolvedMeasureFilters.filter,
+    ),
+    having: sanitiseExpression(undefined, params.criteria),
+    timeRange: {
+      start: timeControls.timeStart,
+      end: timeControls.timeEnd,
+    },
+    limit: "50", // arbitrary limit to make sure we do not pull too much of data
+    sort,
+  };
+}
+
+function getAlertPreviewQueryOptions(
+  ctx: StateManagers,
+  params: AlertPreviewParams,
+  metricsViewSpec: V1MetricsViewSpec,
+): CreateQueryOptions<
+  Awaited<ReturnType<typeof queryServiceMetricsViewAggregation>>,
+  unknown,
+  AlertPreviewResponse
+> {
+  const {
+    selectors: {
+      measureFilters: { getResolvedFilterForMeasureFilters },
+      timeRangeSelectors: { timeControlsState },
+    },
+  } = ctx;
+  const timeControls = get(timeControlsState);
+  // 1st get is to fetch the selector, 2nd get is to get async data.
+  const resolvedMeasureFilters = get(get(getResolvedFilterForMeasureFilters));
```

Comment:
Just a note we won't be able to use this in the Edit Alert dialog where there's no dashboard (and accompanying state) on the page, but let's refactor in a follow-on.